Go flavorful runtime test failure

cc @Mossaka, some example failures are:

https://github.com/bytecodealliance/wit-bindgen/actions/runs/6306187417/job/17120901531
https://github.com/bytecodealliance/wit-bindgen/actions/runs/6305130898/job/17117916779
https://github.com/bytecodealliance/wit-bindgen/actions/runs/6275880667/job/17044416959

where the last one looks like:

```
---- flavorful::run stdout ----
wasi adapter = "/home/runner/work/wit-bindgen/wit-bindgen/target/debug/build/test-artifacts-15b6250480c633e2/out/wasm32-unknown-unknown/release/wasi_snapshot_preview1.wasm"
rust core module = "/home/runner/work/wit-bindgen/wit-bindgen/target/debug/build/test-artifacts-15b6250480c633e2/out/wasm32-wasi/debug/flavorful.wasm"
rust component "/home/runner/work/wit-bindgen/wit-bindgen/target/runtime-tests/flavorful/rust.wasm"
"/home/runner/work/wit-bindgen/wit-bindgen/wasi-sdk-16.0/bin/clang" "--sysroot" "/home/runner/work/wit-bindgen/wit-bindgen/wasi-sdk-16.0/share/wasi-sysroot" "./tests/runtime/flavorful/wasm.c" "/home/runner/work/wit-bindgen/wit-bindgen/target/runtime-tests/flavorful/c-flavorful/flavorful.c" "/home/runner/work/wit-bindgen/wit-bindgen/target/runtime-tests/flavorful/c-flavorful/flavorful_component_type.o" "-I" "/home/runner/work/wit-bindgen/wit-bindgen/target/runtime-tests/flavorful/c-flavorful" "-Wall" "-Wextra" "-Werror" "-Wno-unused-parameter" "-mexec-model=reactor" "-g" "-o" "/home/runner/work/wit-bindgen/wit-bindgen/target/runtime-tests/flavorful/c-flavorful/c-wasm.wasm"
dir: /home/runner/work/wit-bindgen/wit-bindgen/target/runtime-tests/flavorful/go-flavorful
status: exit status: 2
stdout: ------------------------------------------

stderr: ------------------------------------------
panic: runtime error: index out of range [0] with length 0

goroutine 24 [running]:
github.com/tinygo-org/tinygo/compiler.(*builder).createFunctionStart(0xc001772780, 0x0)
	/__w/tinygo/tinygo/compiler/compiler.go:1219 +0x11f8
github.com/tinygo-org/tinygo/compiler.(*builder).createFunction(0xc001772780)
	/__w/tinygo/tinygo/compiler/compiler.go:1308 +0x2c
github.com/tinygo-org/tinygo/compiler.(*compilerContext).getFunction(0xc000158f00, 0xc0016cac00)
	/__w/tinygo/tinygo/compiler/symbol.go:224 +0xdaf
github.com/tinygo-org/tinygo/compiler.(*builder).createFunctionCall(0xc001772600, 0xc00173aac0)
	/__w/tinygo/tinygo/compiler/compiler.go:1883 +0x906
github.com/tinygo-org/tinygo/compiler.(*builder).createExpr(0xc001772600, {0x563ba30, 0xc00173aa80})
	/__w/tinygo/tinygo/compiler/compiler.go:2047 +0x4e5
github.com/tinygo-org/tinygo/compiler.(*builder).createInstruction(0xc001772600, {0x563b910?, 0xc00173aa80?})
	/__w/tinygo/tinygo/compiler/compiler.go:1460 +0xa37
github.com/tinygo-org/tinygo/compiler.(*builder).createFunction(0xc001772600)
	/__w/tinygo/tinygo/compiler/compiler.go:1349 +0xacc
github.com/tinygo-org/tinygo/compiler.(*compilerContext).createPackage(0xc000158f00, {0xc00162d500?}, 0xc00089b200)
	/__w/tinygo/tinygo/compiler/compiler.go:901 +0x8e5
github.com/tinygo-org/tinygo/compiler.CompilePackage({0xc00031e8e8?, 0x5a?}, 0xc0001d5c20, 0xc00089b200, {0x183cabd6d61d5593?}, 0x55a15af?, 0xcd?)
	/__w/tinygo/tinygo/compiler/compiler.go:312 +0x42a
github.com/tinygo-org/tinygo/builder.Build.func3(0xc001644000)
	/__w/tinygo/tinygo/builder/build.go:365 +0x1f4
github.com/tinygo-org/tinygo/builder.runJob(0xc001644000, 0x0?)
	/__w/tinygo/tinygo/builder/jobs.go:222 +0x4d
created by github.com/tinygo-org/tinygo/builder.runJobs in goroutine 1
	/__w/tinygo/tinygo/builder/jobs.go:123 +0x5bd

thread 'flavorful::run' panicked at 'failed to compile', tests/runtime/main.rs:273:13
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace


failures:
    flavorful::run
```

_Originally posted by @alexcrichton in https://github.com/bytecodealliance/wit-bindgen/issues/679#issuecomment-1735657202_

--- a/.github/workflows/main.yml
+++ b/.github/workflows/main.yml
@@ -61,7 +61,7 @@ jobs:
         go-version: '1.20'
     - uses: acifani/setup-tinygo@v1
       with:
-        tinygo-version: 0.29.0
+        tinygo-version: 0.30.0
     - run: cargo test --workspace
     - run: cargo build
     - run: cargo build --no-default-features

--- a/tests/runtime/main.rs
+++ b/tests/runtime/main.rs
@@ -225,7 +225,7 @@ fn tests(name: &str, dir_name: &str) -> Result<Vec<PathBuf>> {
 
     // FIXME: need to fix flaky Go test
     #[cfg(feature = "go")]
-    if !go.is_empty() && name != "flavorful" {
+    if !go.is_empty() {
         let world_name = &resolve.worlds[world].name;
         let out_dir = out_dir.join(format!("go-{}", world_name));
         drop(fs::remove_dir_all(&out_dir));

